ntoskrnl/mm/ARM3/sysldr.c

```diff
@@ -268,28 +256,168 @@ MiLocateExportName(IN PVOID DllBase,
+                            DllBase,
+                            ExportDirectory->NumberOfNames,
+                            NameTable,
+                            OrdinalTable);
+
+    /* Check if we couldn't find it */
+    if (Ordinal == -1)
+        return STATUS_ORDINAL_NOT_FOUND;
+
+    /* Validate the ordinal */
+    if (Ordinal >= ExportDirectory->NumberOfFunctions)
+    {
+        /* It's not, fail */
+        return STATUS_ORDINAL_NOT_FOUND; // STATUS_PROCEDURE_NOT_FOUND;
+    }
 
     /* Resolve the address and write it */
     ExportTable = (PULONG)((ULONG_PTR)DllBase +
                            ExportDirectory->AddressOfFunctions);
-    Function = (PVOID)((ULONG_PTR)DllBase + ExportTable[Ordinal]);
+    FunctionAddress = (PVOID)((ULONG_PTR)DllBase + ExportTable[Ordinal]);
 
-    /* Check if the function is actually a forwarder */
-    if (((ULONG_PTR)Function > (ULONG_PTR)ExportDirectory) &&
-        ((ULONG_PTR)Function < ((ULONG_PTR)ExportDirectory + ExportSize)))
+    if (CheckIfForwarder)
     {
-        /* It is, fail */
-        return NULL;
+        /* Check if the function is actually a forwarder */
+        if (((ULONG_PTR)FunctionAddress > (ULONG_PTR)ExportDirectory) &&
+            ((ULONG_PTR)FunctionAddress < ((ULONG_PTR)ExportDirectory + ExportSize)))
+        {
+            ASSERT((FunctionAddress <= (PVOID)ExportDirectory) ||
+                   (FunctionAddress >= (PVOID)((ULONG_PTR)ExportDirectory + ExportSize)));
+
+            /* It is, fail */
+            return STATUS_PROCEDURE_NOT_FOUND;
+        }
     }
 
     /* We found it */
+    *Function = FunctionAddress;
+    return STATUS_SUCCESS;
+}
+
+/**
+ * @brief
+ * Finds the address in a loaded image of a given exported routine by name.
+ *
+ * @param[in]   DllBase
+ * The base address of the loaded image.
+ *
+ * @param[in]   ExportName
+ * The name of the export, given as an ANSI NULL-terminated string.
+ *
+ * @return
+ * The address of the named exported routine, or NULL if not found.
+ *
+ * @note
+ * This routine was originally named MiLocateExportName(), with a separate
+ * duplicated MiFindExportedRoutineByName() one (taking a PANSI_STRING)
+ * on Windows <= 2003. Both routines have been then merged and renamed
+ * to MiFindExportedRoutineByName() on Windows 8 (taking a PCSTR instead),
+ * and finally renamed and exported as RtlFindExportedRoutineByName() on
+ * Windows 10.
+ **/
+#define MiLocateExportName              RtlFindExportedRoutineByName
+#define MiFindExportedRoutineByName     RtlFindExportedRoutineByName
+PVOID
+NTAPI
+RtlFindExportedRoutineByName(
+    _In_ PVOID DllBase,
+    _In_ PCSTR ExportName)
+{
+    PVOID Function;
+    NTSTATUS Status;
+
+    PAGED_CODE();
+
+    /* Call the internal function */
+    Status = RtlpFindExportedRoutineByName(DllBase,
+                                           ExportName,
+                                           &Function,
+                                           TRUE);
```

Comment:
Why should this fail on a forwarder?

Reply:
The original functions MiFindExportedRoutineByName and MiLocateExportName failed on that case (second one really did, first one kind of asserted, meaning that in release builds this would get unnoticed).

Starting Win8 and Win10 where that exported RtlFindExportedRoutineByName is used in place of those old internal functions, it fails in this condition as well.
I could write a Win10-specific kernel-mode test to show this.
A test that could work on current ROS and Win2k3 would be with MmGetSystemRoutineAddress() (that internally uses this ***FindExportedRoutineByName), to see whether it supports forwarders or not.
But that latter one is limited to looking only at ntoskrnl and hal modules. Also, if an ntos function forwards to a like-named one in hal, the way it is written is that it will succeed, since: a) it will search it in ntos first (and potentially fail if it was a forwarder), then b) search it in hal.

Why this has been done so? One possible reason is that the original internal functions were potentially designed to locate those system routines that must not be forwarders ("DllInitialize", "DllUnload", and all the entry points listed in https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ps/psmgr.c;hb=4d0d22aee22a56b89990c9e18b2a4cb3cf8d8af5#l172).